### DISCUSSION TITLE
How to update an Association Object?

### DISCUSSION THREAD

**Discussion by soerenlorenz** [sqlalchemy/sqlalchemy#9763] [Usage Questions] [answered]
In my use case I have to extensively check for existing objects in the DB by alternate keys. I do this via a customized unique function derived from the [suggestion in the wiki.](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) Now for an Association Object between these objects I thought this might not be necessary as it has a simple PK and I can use `session.merge`. Turns out to be not so easy, which I try to relate to the   [example in the docs](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object):

```python
with Session() as session:
    # father, son, association with extra_data="lost son" already in DB 
    p = _unique(session, Parent, (), {"name": "father",})
    c = _unique(session, Child, (), {"name": "son",})
    # p, c reconciled with DB by _unique function
    a_returned = session.merge(Association(extra_data="lost son, returned", child=c, parent=p))
    # fails, because at merge time the id's are not populated, so instead of an UPDATE for extra_data, an
    # INSERT is emmited
    # a_returned = session.merge(Association(extra_data="lost son, returned", left_id=1, right_id=1))
    # would succeed, but is pointless
    session.commit()
```
What is the preferred way to model Associations that might not yet exist, might already exist or might already exist with outdated extra_data?

**Comment by zzzeek** [comment]
usually you model the Association object so that its primary key is a composite of the identifiers for Parent and Child, so I guess you are doing that, so yes you would normally want to populate the primary key attributes directly.  Why "pointless"?   if that's the row you want to merge, that's the row. if you want p/c associated with it up front, you can do that also:

    a_returned = session.merge(Association(extra_data="lost son, returned", left_id=1, right_id=1, child=c, parent=p))

**Reply by soerenlorenz** [reply]
Yes, PK is the composite of Parent and Child ID's. Pointless, because I can set the ID's manually only in this pet example. In my _real_ code I would not know them or they might not even have ID's yet - as per the result of the `_unique` function.

The docs have an example like this, which handles setting the IDs in the Association automatically:
```python
with Session() as session:
    p = Parent(name="father")
    a = Association(extra_data="lost son")
    c = Child(name="son")
    a.child = c
    p.children.append(a)
    session.add_all([p, a, c])
    session.commit()
```
To rephrase my question: How do I handle the cases where Parent and Child already exist and I want to add an association between them or I want to update the extra_data on their association?

If you say that the association object pattern is not of much help here and I have to collect and set the correct ID's myself, that would be perfectly fine. I just want to make sure not to underutilize your framework.

**Reply by zzzeek** [reply, marked as answer]
> Yes, PK is the composite of Parent and Child ID's. Pointless, because I can set the ID's manually only in this pet example. In my _real_ code I would not know them or they might not even have ID's yet - as per the result of the `_unique` function.

if the unique function returned to you an object that did not yet have an "id", then the id is None and your association row would also have an id of None, so you would not need to set the id.

the use of merge() here implies that you think there's a good chance the Association object is already loaded in the Session.  If it's not already loaded in the Session,  or even if it is but usually it won't be, you can query for it instead, and if it's None, then put your new object in.  that is, do what merge() does one step at a time (see below).  That way you would not need to engineer how to generate its primary key.


> The docs have an example like this, which handles setting the IDs in the Association automatically:
> 
> ```python
> with Session() as session:
>     p = Parent(name="father")
>     a = Association(extra_data="lost son")
>     c = Child(name="son")
>     a.child = c
>     p.children.append(a)
>     session.add_all([p, a, c])
>     session.commit()
> ```

yes that's flushing the object so in the flush process, those ids are set.    it does not happen automatically before then however.

> 
> To rephrase my question: How do I handle the cases where Parent and Child already exist and I want to add an association between them or I want to update the extra_data on their association?

Just make a new Association object and add() it to the Session if one is not present in the database already.   if one might be present, then load it up front and populate it with new data.   it can be loaded given parent and child objects alone, **assuming** those objects are themselves persistent.   

I guess a full series of logic would look like this:


```py

from sqlalchemy import inspect

parent = _unique(...)
child = _unique(...)

# both objects are persistent.  see if there is an Association in the DB

assoc_obj = None
if inspect(parent).persistent and inspect(child).persistent:
    assoc_obj = session.scalar(
          select(Association).where(Association.parent == parent, Association.child == child)
    )

# assoc obj not in DB, either because p/c are not persistent, or couldn't find an Association
if assoc_obj is not None:
    assoc_obj  = Association(parent=parent, child=child, extra_data=extra_data)
    session.add(assoc_obj)
else:
    assoc_obj.extra_data = "data"

session.commit()
```

    
> 
> If you say that the association object pattern is not of much help here and I have to collect and set the correct ID's myself, 

those are two mutually exclusive concepts. the association object pattern is helpful, and if you wanted to use session.merge() for an association object where that object is already present in the DB, you'd have to set its primary key manually.   as above though there are other ways to do this.

**Reply by soerenlorenz** [reply]
That clarifies it for me. Thank you so much for the support and your epic work on the library.

> yes that's flushing the object so in the flush process, those ids are set. it does not happen automatically before then however.

This helps a lot. It seems to be a reoccurring issue of me to assume just to much magic. So I'll just happily cover all the exists-or-not cases step by step in my app code.

Last question: Is that `inspect` function intended to be used in app code or more of a dev tool? What's the advantage of checking for `.persistent` compared to checking just the ID's to be not `None`?

**Reply by zzzeek** [reply]
you could check the ids are `None` also but you seemed to want to abstract away from knowing the primary key column(s).    inspect(obj) gives you the `InstanceState` which can tell you all about the object's status in a way that's agnostic of the table design.

**Reply by zzzeek** [reply]
also technically an object can have primary key that's not None, but the row is not in the database yet; if you make a row with the PK values ready to go.

**Reply by soerenlorenz** [reply]
> inspect(obj) gives you the `InstanceState` which can tell you all about the object's status in a way that's agnostic of the table design.

Ah, obviously. That would be a starting point for more generic implementations.
